src/Servers/Kestrel/shared/test/TestResources.cs

```diff
@@ -13,14 +17,28 @@ public static class TestResources
         public static string TestCertificatePath { get; } = Path.Combine(_baseDir, "testCert.pfx");
         public static string GetCertPath(string name) => Path.Combine(_baseDir, name);
 
-        public static X509Certificate2 GetTestCertificate()
-        {
-            return new X509Certificate2(TestCertificatePath, "testPassword");
-        }
+        private const int MutexTimeout = 120 * 1000;
+        private static readonly Mutex importPfxMutex = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+            new Mutex(initiallyOwned: false, "Global\\KestrelTests.Certificates.LoadPfxCertificate") :
+            null;
 
-        public static X509Certificate2 GetTestCertificate(string certName)
+        public static X509Certificate2 GetTestCertificate(string certName = "testCert.pfx")
         {
-            return new X509Certificate2(GetCertPath(certName), "testPassword");
+            // On Windows, applications should not import PFX files in parallel to avoid a known system-level
+            // race condition bug in native code which can cause crashes/corruption of the certificate state.
+            if (importPfxMutex != null)
+            {
+                Assert.True(importPfxMutex.WaitOne(MutexTimeout), "Cannot acquire the global certificate mutex.");
```

Comment:
Hopefully, this won't cause more starvation right? 😄

Reply:
Famous last words :)

Reply:
The tread-safety issue is related to initialization on Windows, not usage right? I wonder if we should also cache the results for a given certName in a given process. VS is telling me this is called from 33 different places throughout the Kestrel test projects.

That way you wouldn't have to take a system-wide lock for all 33 calls. You could first acquire a normal process-wide lock and see if a cached cert has already been loaded by the process.